Support for logging via Python's logging module.

xavierholt opened the pull request:
#### Reference Issue
Better Python logging support based on the discussion from #4840.

#### What does this implement/fix?
This adds a few new logging options that make controlling RDKit logs from Python easier.  The big changes are the following new methods, all of which _replace_ any previously-initialized log streams:

- `rdkit.Chem.LogToCppStreams()`  This simply exposes the default `RDLog::InitLogs()` via the Python API.  This can be useful because most of the logging functions will silently fail if called before the log streams are initialized, and there's currently no (obvious) way to initialize them.
- `rdkit.Chem.LogToPythonStderr()`  This initializes the RDKit logs so that they print via Python's `sys.stderr` stream.  This works a lot like `rdkit.Chem.WrapLogs()`, but without the tee and without the `RDKit LEVEL: ` prefix.
- `rdkit.Chem.LogToPythonLogger()`  This initializes the RDKit logs so that they print using Python's `logging` module (specifically via the `logging.getLogger("rdkit")` logger) at the appropriate log levels.  If you want access to the RDKit logs in Python, this is most likely the init function you want.

I also added `LogDebugMsg` and `LogInfoMsg` functions to allow access to those streams from Python (useful for testing), and a logging test suite.

#### Any other comments?
A couple things I was wondering about:
- Is there a way to detect whether we're running in C++ or Python?  If so, we could initialize the logs based on that.
- The C++ info log writes to `std::cout`, but all the others write to `std::cerr`.  Is this intentional?
- It seems a little weird to have logging control as part of the `Chem` module.  Is there a better place for this?

greglandrum commented:
@xavierholt : thanks for this contribution! It's awesome to be able to use the standard Python logging stuff to control how verbose the RDKit is and I think this will make a lot of folks happy.

The changes look mostly great, but I would really prefer to have this new functionality in `rdkit.rdBase`, not `rdkit.Chem`. The fact that there is currently logging functionality in `Chem` is an unfortunate historical thing which we should slowly fix, but there's no reason to dig the hole deeper. Could you please move the new stuff to rdBase and the test (thanks for doing some!) to `$RDBASE/rdkit`?

bp-kelley commented on Code/GraphMol/Wrap/rdchem.cpp:
```diff
@@ -78,28 +79,100 @@ struct PySysErrWrite : std::ostream, std::streambuf {
-  static PySysErrWrite error("RDKit ERROR: ");
-  static PySysErrWrite info("RDKit INFO: ");
-  static PySysErrWrite warning("RDKit WARNING: ");
+  static PyErrStream debug("RDKit DEBUG: ");
```
The intent of WrapLogs was to be a lot closer to what LogToPythonLogger does.  Also windows systems have a flushing issue with WrapLogs that may be fixed with LogToPythonLogger.

I expect WrapLogs should just call LogToPythonLogger.

xavierholt replied:
Okay!  My original plan was to leave all the existing functions untouched, but I agree that `LogToPythonLogger` is more useful / less surprising than the old `WrapLogs` behavior.  I have a commit coming soon that'll address this.

Do you have a link for the Windows flushing problems?  I doubt my changes will fix anything, 'cause they use the same underlying stream implementation, but I can take a look while I'm working on it.

greglandrum replied:
I think it'd be fine to postpone the windows problems to a separate PR

bp-kelley commented on Code/GraphMol/Wrap/rdchem.cpp:
```diff
@@ -78,28 +79,100 @@ struct PySysErrWrite : std::ostream, std::streambuf {
+}
+
+void RDLogInfo(const std::string &msg) {
+  NOGIL gil;
```
I note that we don't have any threading tests which make me a bit nervous.

Perhaps we should dispense with the NOGIL entirely here as these will always be called via python and the expectation is that they will write to a python stream.

xavierholt replied:
I'm unsure about this one - my current version is just copying what was already implemented for errors and warnings.  There is a possible case where some Python code is using these functions to log via the C++ logger.  Not sure what the performance impacts would be - whatever the overhead is to release/reclaim the GIL vs. whatever the overhead is in dealing with `std::ostream`s (I'm guessing it's about even, but have zero data to back that up).

I'll definitely add some threading tests, though!

greglandrum replied:
I'd prefer to deal with changing the GIL situation (if necessary) in a separate PR.
This one already has useful functionality and diving into the rabbit hole which is cross-language thread safety and logging here seems like a bad idea

bp-kelley commented:
I think logs should always write to stderr, but greg can make the call here.

bp-kelley commented:
@xavierholt we have a bit of code in rdkit/__init__.py that checks to see if we are running jupyter and automatically wraps the logs:

https://github.com/rdkit/rdkit/blob/d20077a089ca9e79b72a41934ac1d74a19bec3c3/rdkit/__init__.py

We did this for jupyter because otherwise you can't really see ANY rdkit logging output.  I personally don't mind making this the default behavior, but it has the potential to have a speed impact for some systems and should be reversible if we go down this path.

xavierholt commented:
> I think logs should always write to stderr, but greg can make the call here.

I'm in complete agreement on this one.  Being able to run `process data.in > data.out` without having to worry about getting log messages in your data file is extremely convenient.  @greglandrum what are your thoughts on this?

greglandrum commented:
> I think logs should always write to stderr, but greg can make the call here.

I think there's an argument to be made that INFO can go somewhere else, but I'm not dogmatic about that.

A change in behavior could break existing workflows for people, so if we want to change where INFO goes by default we'll need to deprecate the behavior in one release and make the change in the next one, with a way to get the old behavior back for people who want/need it.

I'm fine with this *new* code changing things around (i.e. if you tell it to use the python logger then it goes wherever the python logger goes by default). I just want to be careful about changing the behavior of existing scripts.

greglandrum commented:
> > I think logs should always write to stderr, but greg can make the call here.
> 
> I'm in complete agreement on this one. Being able to run `process data.in > data.out` without having to worry about getting log messages in your data file is extremely convenient. @greglandrum what are your thoughts on this?

I totally agree that it's great to be able to do that. And if we can make it easy/possible to have that happen with new code I think it's great. As I mentioned in the previous commit: I don't think we can just change the default behavior though. We'd need to do that over a couple of releases.

greglandrum commented on rdkit/Chem/test_list.py:
```diff
@@ -1,46 +1,47 @@
-         }), ("python", "test_list.py", {
-           'dir': 'MolStandardize'
-         })]
+tests = [
```
Reordering the tests here isn't really part of the change you're trying to make and doesn't add anything. Please revert this.

xavierholt commented:
Thanks for the feedback!  I've got a new version in progress that should take care of most of the requested changes, and I can open a separate issue for the info log location (ideally I'll change as little as possible in this PR).

Unfortunately, I've run into some GIL/threading issues that show up in the Python tests when you use the Python logger by default - I'm still working on it, but it might take me another day or two to figure out exactly what's going on.

greglandrum commented:
> Thanks for the feedback! I've got a new version in progress that should take care of most of the requested changes, and I can open a separate issue for the info log location (ideally I'll change as little as possible in this PR).

awesome, thanks!

> Unfortunately, I've run into some GIL/threading issues that show up in the Python tests when you use the Python logger by default - I'm still working on it, but it might take me another day or two to figure out exactly what's going on.

I admit that the whole GIL thing confuses the hell out of me, but I believe that it may be a mistake to be using NOGIL in this code since the C++ code actually calls back into Python when you're using the python logger.

I believe that it's fine to release the GIL as long as you're in (threadsafe) pure C++ code, like all the existing logging code, but in your case it's probably wrong.

bp-kelley commented:
My point was that we are now releasing the GIL and potentially calling into the python API which may try to reacquire it.  I think the "safe approach" is to simply not release the GIL when doing python logging.

bp-kelley commented:
@xavierholt if you push the failing tests to the repo (you might have already) I can help take a look.

xavierholt commented on Code/GraphMol/FilterCatalog/Wrap/FilterCatalog.cpp:
```diff
@@ -299,6 +299,13 @@ python::dict GetFlattenedFunctionalGroupHierarchyHelper(bool normalize) {
   return dict;
 }
+
+std::vector<std::vector<boost::shared_ptr<const FilterCatalogEntry>>>
+RunFilterCatalogWrapper(const FilterCatalog &fc, const std::vector<std::string> &smiles, int numThreads) {
+  NOGIL nogil;
+  return RunFilterCatalog(fc, smiles, numThreads);
+}
```
This scares me.  I had to add this wrapper to avoid a deadlock in the tests - I think creating new threads while holding the GIL messed it up somehow, and then when the Python logger tried to grab the GIL it wasn't able to, and locked up hard (I couldn't even `Ctrl+C` - had to `Ctrl-\`).

Adding the `NOGIL` wrapper fixed it, but my concern is that there are other functions that:
- Create threads while holding the GIL, and
- Aren't tested with invalid inputs, and therefore don't log.

Anything that meets those criteria runs a risk of hard lockup in production (when using either Python logger).

bp-kelley replied:
It bothers me as well.  I'll take a look.

greglandrum replied:
I believe this is now resolved, right?

xavierholt replied:
Should be.  I pulled in Brian's fixes for Filter Catalog and Substruct Library, and I think those were the only places that had a similar dangerous pattern.

xavierholt commented on Code/RDGeneral/RDLog.cpp:
```diff
@@ -120,13 +120,12 @@ void InitLogs() {
-          << std::setw(2) << std::setfill('0') << details.tm_min << ":"
-          << std::setw(2) << std::setfill('0') << int(details.tm_sec) << "] ";
-  return logstrm;
+  strftime(buffer, 16, "[%T] ", localtime(&t));
```
This gives significantly cleaner output in multithreaded situations.  It's not perfect, but it's much better.

xavierholt commented on rdkit/Chem/__init__.py:
```diff
@@ -163,6 +163,18 @@ def FindMolChiralCenters(mol, force=True, includeUnassigned=False, includeCIP=Tr
   return centers
 
+import warnings
+def WrapLogs():
```
Make sure the old logging functions still work, but add deprecation warnings (unfortunately Python suppresses deprecation warnings by default, so I'm not sure if anyone'll notice).

greglandrum replied:
Can you please add something to ReleaseNotes.md for this release for this?
The section header is:
```
## Deprecated code (to be removed in a future release):
```

xavierholt replied:
Sure!  Just added the deprecation notes.  Should I also add the new logging control functions to the Highlights section?  I'll call them experimental, in case there are threading bugs we haven't found yet.

xavierholt commented on rdkit/__init__.py:
```diff
@@ -7,17 +11,23 @@
+
+# Do logging setup at the end, so users can suppress the
+# "enabling jupyter" message at the root logger.
+log_handler = logging.StreamHandler(sys.stderr)
+logger.addHandler(log_handler)
+logger.setLevel(logging.DEBUG)
+logger.propagate = False
```
This makes the Python logger output all levels, to match the behavior of the C++ and Python `stderr` loggers.  It's a little weird to add a handler to a non-root logger like this, but I figured it was better than changing people's default logging setup whenever they `import rdkit`.

xavierholt commented on Code/RDBoost/Wrap/RDBase.cpp:
```diff
@@ -48,33 +194,41 @@ void AttachFileToLog(std::string spec, std::string filename, int delay = 100) {
 }
+
+
+void LogDebugMsg(const std::string &msg) {
+  NOGIL nogil;
+  BOOST_LOG(rdDebugLog) << msg << std::endl;
+}
```
These are still releasing the GIL for now, 'cause it's the old behavior, but the tests run fine without this as well.

bp-kelley replied:
The only reason they released the GIL was because C++ logging doesn't need it.

bp-kelley commented on rdkit/__init__.py:
```diff
@@ -7,17 +11,23 @@
   pass
+
+# Do logging setup at the end, so users can suppress the
+# "enabling jupyter" message at the root logger.
```
I don't think we should use the python logger by default.

xavierholt replied:
That makes sense (it's the least invasive).  Should we use it for Jupyter, though?  The old version used `WrapLogs()` if it detected Jupyter.

greglandrum replied:
I'd rather not make any changes to default behavior as part of this PR.
If we're going to do that, let's discuss it in a separate PR

xavierholt replied:
Sounds good.  I'd like to leave it enabled for now, though, so we can make sure all the tests can pass CI with Python logging enabled - once we know it's good I'll revert to the old behavior.  (Unless there's a better way to do this?)

greglandrum commented:
@xavierholt : if you don't have a windows machine to test/debug this on directly, let me know. I can spend some time with it and see if I can figure out what's going on.

xavierholt commented:
> @xavierholt : if you don't have a windows machine to test/debug this on directly, let me know. I can spend some time with it and see if I can figure out what's going on.

I'd appreciate it!  I only have access to a Mac at the moment (and Linux via Docker).  The first failing commit added multi-threading tests to `rdkit/UnitTestLogging.py`, and I'm pretty sure there's some sort of lock-up in (at least) one of those - I'd check out the `testAsynchronousN()` tests first.  If things are working right, that whole test file should run in under a second.

greglandrum commented:
> 
> I'd appreciate it! I only have access to a Mac at the moment (and Linux via Docker). The first failing commit added multi-threading tests to `rdkit/UnitTestLogging.py`, and I'm pretty sure there's some sort of lock-up in (at least) one of those - I'd check out the `testAsynchronousN()` tests first. If things are working right, that whole test file should run in under a second.

Ok, I will (try to) make some time to take a look today.

bp-kelley commented:
One of the issues I think I'm seeing is the following:

```
python -> calls rdkit function potentially holds GIL
   C++ -> indirectly calls python function, maybe via logging or maybe via a callback like Filtercatalog
         GIL is now recursively locked and we're hosed.
```

I'm wondering if we should wrap the writer in something like:

```
if (PyGILState_Check()) {
    /* I have the GIL so don't lock it twice */
} else {
    /* ok to lock the GIL */
}
```

bp-kelley commented:
I'm admittedly now having trouble making a locking scenario, but I'm working on it.

bp-kelley commented:
I have a fix for the filter catalog, but it's a bit invasive.

Essentially we need to lock the GIL per-thread.

RunFilterCatalog - releases the GIL
  Then calls a search callback which must lock the GIL so we can log and do other pythonic stuff.

I think that this will be required for all of our threaded calls that can call back into python.

I think this fix is required for the FilterCatalog anyway since we already have python callbacks, I'll push this as another PR for inspection.

xavierholt commented:
@bp-kelley I don't think we need to do any extra checking to avoid locking the GIL twice.  The [docs](https://docs.python.org/3/c-api/init.html#c.PyGILState_Ensure) for `PyGILState_Ensure()` (which is what the `PyGILStateHolder` class uses) say:

> This may be called as many times as desired by a thread as long as each call is matched with a call to `PyGILState_Release()`.

Hopefully that simplifies things.

bp-kelley commented:
@xavierholt 

The issue here is threading.  If the GIL is held in the main thread and then stuff happens in a worker thread I am seeing a lot of crashing issues.

the key wording here is:

```
as many times as desired by a thread 
```

greglandrum commented:
> 
> I'd appreciate it! I only have access to a Mac at the moment (and Linux via Docker). The first failing commit added multi-threading tests to `rdkit/UnitTestLogging.py`, and I'm pretty sure there's some sort of lock-up in (at least) one of those - I'd check out the `testAsynchronousN()` tests first. If things are working right, that whole test file should run in under a second.

For what it's worth: the Windows problems have nothing whatsoever to do with the new code, which works fine. The problems are being caused by `CaptureStream` in the testing code. I will see if I can come up with a solution for that today.

greglandrum commented:
> 
> For what it's worth: the Windows problems have nothing whatsoever to do with the new code, which works fine. The problems are being caused by `CaptureStream` in the testing code. I will see if I can come up with a solution for that today.

The hangs which are occurring during the windows tests seem to be related to a write buffer size in the object which is returned by `os.pipe()`. It's not connected to threading or even the new C++ code that's part of this PR: I can reproduce the hang using pure python code (with no calls into the RDKit at all).

Reducing the number of log entries created on Windows resolves the problem (at least for me). It seems to me that this is the way to go and that digging into the root causes to get the tests shiny and perfect on Windows is a waste of time, particularly since the effort has nothing whatsoever to do with the RDKit itself.

I'll do a PR against @xavierholt's branch with the suggested changes to the tests.

bp-kelley commented:
@greglandrum that's good to know.  The FilterCatalog issue is real though, but I can handle that in a separate PR after or during this one.  I'll also check the substruct library as it can be pretty chatty with error messages during smiles parsing and uses the same threading protocol.

greglandrum commented:
> @greglandrum that's good to know. The FilterCatalog issue is real though, but I can handle that in a separate PR after or during this one. I'll also check the substruct library as it can be pretty chatty with error messages during smiles parsing and uses the same threading protocol.

Yeah, I'm sure that the fact that the C++ code is multithreaded and can call back into Python opens the doors to a variety of painful problems, I was just really glad that it's not a factor for this PR.

I think this is a high-value feature add and I'd really like to get it in there. :-)

xavierholt commented:
This is good news!  I was worried I broke Windows there...  If you let me know safe values of `nthreads` and `nlogs` I can make a quick commit and you won't have to bother making a PR.

greglandrum commented:
> This is good news! I was worried I broke Windows there... If you let me know safe values of `nthreads` and `nlogs` I can make a quick commit and you won't have to bother making a PR.

I did the PR earlier today

xavierholt commented:
@greglandrum - Oh, thanks!  For some reason I didn't get a notification...  Merged now.

bp-kelley commented:
Nice.  I think this is mergeable as long as we don't make it the default for now.

xavierholt commented:
A couple notes on those last commits:
- The default logger is back to C++, as it was originally.
- I changed the threaded logging tests to capture output in a tempfile rather than a pipe, and they can handle the full run on Windows now too.
- The two commits that failed CI both had errors like this:

```
FAIL: testAsynchronous2 (__main__.TestLogToPythonLogger)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "UnitTestLogging.py", line 322, in testAsynchronous2
    self.assertEqual(captured['DEBUG'  ].count('Debug'),   nthreads * nlogs)
AssertionError: 272 != 250
```

It only seems to happen on that test on the Ubuntu / minimal dependencies build.  I'll take another look and see if I can find anything weird, and I've added a print statement to the test so we'll get more info if it fails in CI again.

greglandrum commented on Code/GraphMol/SubstructLibrary/Wrap/rough_test.py:
```diff
@@ -616,6 +616,50 @@ def testPropHolder(self):
       self.assertEqual(['Z11234'], list(library.GetKeyHolder().GetKeys(indices)))
 
-
+  def test_bad_smiles(self):
```
This is really horribly verbose when it runs and fills the console in a way that makes it hard to figure out what's going on.
Any chance we can reduce the apparent verbosity of this by using the Python logger and then disabling that log? That would still make all the logging calls desired but wouldn't pollute the logs quite so badly.

bp-kelley commented:
@greglandrum I'll see if we can disable the python log in both tests.  They will still throw C++ logs though I think.   I can also make them a separate test so at least they can be separated easily if we need to.

greglandrum commented:
@xavierholt : I just did a PR against your branch to suppress the log messages from the recent changes by @bp-kelley. If you're ok with those, I think this PR is ready to merge.

xavierholt commented:
Thanks!  I think it's ready now too.

greglandrum commented:
@xavierholt : thanks for the contribution. This will be really useful for people building production tools on top of the RDKit.
